### ISSUE TITLE
How the reverse proxy does handle requests?

### ISSUE THREAD

**Issue by rlevchenko** [microsoft/service-fabric#435]
### Scenario

When I hit my services through the localhost:19081/applicationPath it seems like that the Reverse Proxy tries to forward requests to the remote node as well (I assume). If we have NSGs configured to deny one of the services port, for instance, such requests may ended up with timeout issue (see below) or the response becomes very slow (up to 2 mins, but localhost:servicePort responses instantly). Workaround: make sure that services traffic is not denied between the node types..

So, does the reverse proxy also act as an internal load balancer? Sorry for may be silly question, but I haven't found any detailed description of the reverse proxy flow/architecture.

### Error
“Forwarding request to service failed, retrying with re-resolve: phase = SendRequestHeaders, error = -805,306,139, internal error = ERROR_WINHTTP_TIMEOUT ") I guess it happens when the reverse proxy forwards the request to the remote node and, when the timeout is reached, redirects the request to the localhost (finally) and service web app opens. A bit strange..but I don't have other explanation why NSGs  have impact here...I thought localhost is localhost..

### Environment

- Azure, Service Fabric Cluster
- 2 node types (seed/secondary)
- 6 stateless applications
- Reverse Proxy (HTTP)

**Comment by rlevchenko** [comment]
Anyone please? The question is quiet simple..

**Comment by ammarZaidi** [comment]
Roman,
a quick rundown on basic reverse proxy request scenario:
When you use reverse proxy, it gets the registered endpoints for the target service and tries to send the request to one of those endpoints. 
Reverse Proxy has a timeout for each attempt set to 5 secs, it retries to send request for 120 secs before timing out on request.
When a service A tries to connect to a service B using reverse proxy, the endpoint looks something like http://<nodeip>:<service_port>. When you try to connect to a service using nodeIp instead of localhost, the request goes out of the node as if the destination was some other machine. In that scenario, NSG comes into picture and if you have the service port denied on the NSG, the request will not reach the service.
In this case Reverse Proxy will try to re-resolve the service endpoint thinking that the endpoint used earlier is stale but since the port is blocked it will eventually timeout after 120 secs.
In short, if you access a service using IP(reverse proxy will always have IP in the endpoint), you will have to open the port in NSG. 
I hope it answers your question.
